contracts/upgrader/src/contract.rs

```diff
@@ -0,0 +1,138 @@
+#![allow(dead_code)]
+
+use crate::error::ContractError;
+use axelar_soroban_std::traits::ThenOk;
+use soroban_sdk::{contract, contractimpl, Address, BytesN, Env, String, Symbol, Val};
+
+const MIGRATE: &str = "migrate";
+const UPGRADE: &str = "upgrade";
+const VERSION: &str = "version";
+
+#[contract]
+pub struct Upgrader;
+
+#[contractimpl]
+impl Upgrader {
+    pub fn __constructor(_env: Env) {}
+
+    pub fn upgrade(
+        env: Env,
+        contract_address: Address,
+        new_version: String,
+        new_wasm_hash: BytesN<32>,
+        migration_data: soroban_sdk::Vec<Val>,
+    ) -> Result<(), ContractError> {
+        ensure_new_version_is_different(&env, &contract_address, &new_version)?;
+
+        env.invoke_contract::<()>(
+            &contract_address,
+            &Symbol::new(&env, UPGRADE),
+            soroban_sdk::vec![&env, new_wasm_hash.into()],
+        );
+        env.invoke_contract::<()>(
+            &contract_address,
+            &Symbol::new(&env, MIGRATE),
+            migration_data,
+        );
+
+        ensure_new_version_matches_expected(&env, &contract_address, &new_version)
+    }
+}
+
+fn ensure_new_version_is_different(
+    env: &Env,
+    contract_address: &Address,
+    new_version: &String,
+) -> Result<(), ContractError> {
+    let string = current_version(env, contract_address);
+    let no_match = string != *new_version;
+    no_match.then_ok((), ContractError::SameVersion)
+}
+
+fn ensure_new_version_matches_expected(
+    env: &Env,
+    contract_address: &Address,
+    new_version: &String,
+) -> Result<(), ContractError> {
+    let versions_match = current_version(env, contract_address) == *new_version;
+    versions_match.then_ok((), ContractError::VersionMismatch)
+}
```

Comment:
`ensure_new_version_is_different` and `ensure_new_version_matches_expected` have similar logic but different number of lines: 3 vs 2 lines.

Reply:
good catch, the `string` variable was a leftover from some debugging